OCPBUGS-45895: Reconcile pre-existing Secrets

rexagod opened the pull request:
Until now, CMO did not reconcile existing secrets, even if their data changed. This changes that behavior.

<!--
    Don't forget about CHANGELOG if this affects the end user!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Monitoring <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR
    <Component> Component affected by your changes such as deps bump, alerts changes and any user facing changes.

    Example:
    - [#741](https://github.com/openshift/cluster-monitoring-operator/pull/741) Bump thanos components to v0.11.0 release
-->

* [ ] I added CHANGELOG entry for this change.
* [x] No user facing changes, so no entry in CHANGELOG was needed.

machine424 commented on pkg/tasks/alertmanager.go:
```diff
@@ -77,7 +77,7 @@ func (t *AlertmanagerTask) create(ctx context.Context) error {
 	}
 
-	err = t.client.CreateIfNotExistSecret(ctx, s)
+	err = t.client.CreateOrUpdateSecret(ctx, s)
```
Good catch for the rbac proxy secrets, we should for sure update them when needed. Not sure about how we can easily have a regression test for this.

For some secrets like the one containing alertmanager config, CMO shouldn't override it as it may be customized by users https://docs.openshift.com/container-platform/4.16/observability/monitoring/managing-alerts.html#configuring-notifications-for-default-platform-alerts_managing-alerts

rexagod replied:
Yeah, that's a tricky case, since if we make updates to that secret between releases, and upgraded CMO version won't reconcile it.

rexagod replied:
I'll revert the change for those cases.

machine424 replied:
we can go through them one by one ;)
But AFAIK, users aren't meant to change the rbac proxy secrets.

rexagod replied:
Refactored to exclude `alertmanager` configuration secrets for UWM and in-cluster environments, as users may want to modify those.

machine424 replied:
I have an idea of an e2e test for this:
- keep a list of secretsNotToSync (with the alertmanager config secret etc, the first test run will help us build it)
- loop over all the secrets in the two monitoring namespaces (if we can get all the operands enabled it'd be great) if not in secretsNotToSync, alter them (add a `foo: bar` data entry or sth)
- trigger a CMO sync and make sure the added entry gets deleted from all secrets.

We can extend this to configmaps, RBAC etc. in the future.
If we can have it for secrets as part of this PR it'd be great.

rexagod replied:
Sounds good, patched.

rexagod commented:
/jira refresh

rexagod committed:
OCPBUGS-45895: allow reconciling existing Secrets

Until now, CMO did not reconcile existing secrets, even if their data
changed. This changes that behavior.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod committed:
chore: add tests for secret updation behavior

rexagod commented:
I'm currently having some problems with `make versions generate` on MacOS.

rexagod commented:
/retest-required

machine424 commented on Documentation/deps-versions.md:
```diff
@@ -1,7 +1,7 @@
 |--------------|----------------------------------------------------------------------------------|--------------------------------------------------------------------------|-----------------------------------------------------------------------------|-----------------------------------------------------------------------------------|--------------------------------------------------------------- ...
-| release-4.20 | [0.27.0](https://github.com/openshift/prometheus-alertmanager/blob/release-4.20) | [0.18.2](https://github.com/openshift/kube-rbac-proxy/blob/release-4.20) | [2.13.0](https://github.com/openshift/kube-state-metrics/blob/release-4.20) | [0.7.2](https://github.com/openshift/kubernetes-metrics-server/blob/release-4.20) | [1.0.0](https://github.com/openshift/monitoring-plugin/blob/re ...
-| release-4.19 | [0.27.0](https://github.com/openshift/prometheus-alertmanager/blob/release-4.19) | [0.18.2](https://github.com/openshift/kube-rbac-proxy/blob/release-4.19) | [2.13.0](https://github.com/openshift/kube-state-metrics/blob/release-4.19) | [0.7.2](https://github.com/openshift/kubernetes-metrics-server/blob/release-4.19) | [1.0.0](https://github.com/openshift/monitoring-plugin/blob/re ...
+| release-4.20 | [0.27.0](https://github.com/openshift/prometheus-alertmanager/blob/release-4.20) | [0.18.2](https://github.com/openshift/kube-rbac-proxy/blob/release-4.20) | [2.13.0](https://github.com/openshift/kube-state-metrics/blob/release-4.20) | [0.7.2](https://github.com/openshift/kubernetes-metrics-server/blob/release-4.20) | [1.0.0](https://github.com/openshift/monitoring-plugin/blob/re ...
```
If you could rebase, those changes are already on `main`

machine424 commented on test/e2e/reconcile_objects_test.go:
```diff
@@ -0,0 +1,185 @@
+		extractNamespacedNameFromFileSecret = extractNamespacedNameFromFile[*v1.Secret]
+	)
+	unsyncedSecrets := []types.NamespacedName{
+		extractNamespacedNameFromFileSecret(t, manifests.AlertmanagerConfig),
```
I think it's simpler/more readable/less error prone is we just hardcode the names here

machine424 replied:
we don't add new secrets regularly, it's always better to minimize risks that the test itself is broken.
Having hardcoded names is more readable (no need to go refer to the manifest to get the name of a secret) and we don't have to worry that the "fetch secret code" is broken (I'm not saying it's broken now :))

We can always change in the future if we start adding/removing secrets regularly in the future.

machine424 commented on test/e2e/reconcile_objects_test.go:
```diff
@@ -0,0 +1,185 @@
+		extractNamespacedNameFromFileSecret(t, manifests.AlertmanagerConfig),
+		extractNamespacedNameFromFileSecret(t, manifests.AlertmanagerUserWorkloadSecret),
+	}
+	syncedSecrets := []types.NamespacedName{
```
As I see it, we can do without this (to avoid the maintenance burden also)

We can do with:
list all secrets in the namespace, ensure there are at least 5 (we know there are more than 5 secrets) or `len(unsyncedSecrets)` or sth, to make sure the test is really doing sth.

Then loop over all the secrets in there and alter them.

Then loop over all of them, if they're not in `unsyncedSecrets`, check the suffix was removed.

Then at the end, loop over `unsyncedSecrets` and make sure the prefix is still there. (this way we can say we waited for all the syncs to get triggered and those secrets weren't touched)

We can have subtests or sth to run the test on both namespaces or you can consider the secrets frpm both of them in one shot

rexagod replied:
Nice idea, let me patch this.

rexagod replied:
Thinking a bit more on this, let's say if a `synced` secret unintentionally becomes `unsynced`, we'd be only able to catch it randomly?

machine424 replied:
if a `synced` secret becomes `unsynced` the step where we'd check that the suffix is removed will fail.

machine424 reviewed (commented):
thanks for the test, some suggestions.

juzhao commented:
/label qe-approved

rexagod commented:
@machine424 Made the requested changes, PTAL :)

rexagod commented:
`okd-scos-e2e-aws-ovn` errors are unrelated, so this is good for reviews.

machine424 commented on test/e2e/reconcile_objects_test.go:
```diff
@@ -0,0 +1,168 @@
+		for _, secret := range unsyncedSecrets {
+			gotSecret, err := f.KubeClient.CoreV1().Secrets(secret.Namespace).Get(context.Background(), secret.Name, metav1.GetOptions{})
+			if err != nil {
+				if errors.IsNotFound(err) {
```
I think we should be made aware of this; whether `unsyncedSecrets` would need to be updated or the test deletes the secret by mistakes in both cases we need to surface that

machine424 replied:
a log can be missed, let's just remove this special treatment for `IsNotFound` and just fail

machine424 commented on test/e2e/reconcile_objects_test.go:
```diff
@@ -0,0 +1,168 @@
+	// Prepare synced secrets.
+	var syncedSecrets []types.NamespacedName
+	secretsNS, err := f.KubeClient.CoreV1().Secrets(f.Ns).List(context.Background(), metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/managed-by=cluster-monitoring-operator",
```
because the two namespaces are our responsibility, let's just fetch all; if another test leaves a secret around we need to make it clean it, if another operator doesn't sync a secret that it should, we want to identify that and reach out to them

machine424 commented on Documentation/deps-versions.md:
```diff
@@ -1,8 +1,8 @@
-| release-4.20 | [0.27.0](https://github.com/openshift/prometheus-alertmanager/blob/release-4.20) | [0.18.2](https://github.com/openshift/kube-rbac-proxy/blob/release-4.20) | [2.13.0](https://github.com/openshift/kube-state-metrics/blob/release-4.20) | [0.7.2](https://github.com/openshift/kubernetes-metrics-server/blob/release-4.20) | [1.0.0](https://github.com/openshift/monitoring-plugin/blob/re ...
-| release-4.19 | [0.27.0](https://github.com/openshift/prometheus-alertmanager/blob/release-4.19) | [0.18.2](https://github.com/openshift/kube-rbac-proxy/blob/release-4.19) | [2.13.0](https://github.com/openshift/kube-state-metrics/blob/release-4.19) | [0.7.2](https://github.com/openshift/kubernetes-metrics-server/blob/release-4.19) | [1.0.0](https://github.com/openshift/monitoring-plugin/blob/re ...
-| release-4.18 | [0.27.0](https://github.com/openshift/prometheus-alertmanager/blob/release-4.18) | [0.18.1](https://github.com/openshift/kube-rbac-proxy/blob/release-4.18) | [2.13.0](https://github.com/openshift/kube-state-metrics/blob/release-4.18) | [0.7.2](https://github.com/openshift/kubernetes-metrics-server/blob/release-4.18) | [1.0.0](https://github.com/openshift/monitoring-plugin/blob/re ...
+| release-4.20 | [0.28.0](https://github.com/openshift/prometheus-alertmanager/blob/release-4.20) | [0.19.0](https://github.com/openshift/kube-rbac-proxy/blob/release-4.20) | [2.15.0](https://github.com/openshift/kube-state-metrics/blob/release-4.20) | [0.7.2](https://github.com/openshift/kubernetes-metrics-server/blob/release-4.20) | [1.0.0](https://github.com/openshift/monitoring-plugin/blob/re ...
```
if we could have those versions bumps in another PR, that'd be great

machine424 reviewed (commented):
Thanks again, we're almost ready to merge this, just some final suggestions

rexagod commented:
🤦🏼 So I just noticed I had a patch on local for this that I never pushed, I did that now, and I think some of the review comments are addressed by it. I'll go over them again to make sure I act on them.

rexagod commented on test/e2e/reconcile_objects_test.go:
```diff
@@ -0,0 +1,168 @@
+`)
+	f.MustCreateOrUpdateConfigMap(t, userWorkloadConfigMap)
+	defer f.MustDeleteConfigMap(t, userWorkloadConfigMap)
+	for _, secret := range []types.NamespacedName{
```
Used a `require.NotEmpty(t, syncedSecrets)` right after we fill `syncedSecrets` to ensure it's not empty.

machine424 replied:
then I don't see why this `f.AssertSecretExists` is needed.

rexagod commented on test/e2e/reconcile_objects_test.go:
```diff
@@ -0,0 +1,168 @@
+	require.NotEmpty(t, syncedSecrets)
+
+	// Update the aforementioned secrets' data.
+	for _, secret := range append(syncedSecrets, unsyncedSecrets...) {
```
Not sure what this means, could you please elaborate a bit? There's no `secrets` IIUC but only `syncedSecrets` and `unsyncedSecrets`.

machine424 replied:
I meant we can only have `secrets` with all secrets and loop over them only once and then use the `slices.Contains`  check to see whether we want to `// Check if the secrets were reconciled as expected.` or `// Check if the secrets were reconciled unexpectedly.`. so we only have one check loop.
(this is a nit, I just find it more readable that way)

machine424 commented on test/e2e/reconcile_objects_test.go:
```diff
@@ -0,0 +1,151 @@
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func TestSecretsReconciliation(t *testing.T) {
```
let's add a short comment about the overall logic in the test

machine424 commented on test/e2e/reconcile_objects_test.go:
```diff
@@ -0,0 +1,151 @@
+func TestSecretsReconciliation(t *testing.T) {
+	// Create assets under both scenarios for us to work with.
+	setupUserWorkloadAssetsWithTeardownHook(t, f)
+	userWorkloadConfigMap := f.BuildUserWorkloadConfigMap(t, `alertmanager:
```
good catch, I think this config should make CMO produce all/most of the secrets.

rexagod commented:
Need to test this out on a cluster, will request a review once that's green.

rexagod commented:
/retest

juzhao commented:
/retest

machine424 commented:
/retest-required

rexagod commented:
This needs to be updated a bit, will request a review once I clean this up.

machine424 reviewed (commented):
I realized I never pushed the comments

rexagod commented on test/e2e/reconcile_objects_test.go:
```diff
@@ -0,0 +1,170 @@
+				data[k] = []byte(t.Name() + string(v))
+			}
+			break
+		}
```
I believe this is addressed now: https://github.com/openshift/cluster-monitoring-operator/pull/2539/changes#diff-940ff24a237421831783d2b59569062d3a03444707cc4fe8b990c99a6f3d09d9R156

rexagod committed:
fixup! chore: add tests for secret updation behavior

rexagod committed:
fixup! fixup! chore: add tests for secret updation behavior

rexagod committed:
fixup! fixup! fixup! chore: add tests for secret updation behavior

rexagod committed:
fixup! fixup! fixup! fixup! chore: add tests for secret updation behavior

rexagod committed:
fixup! fixup! fixup! fixup! fixup! chore: add tests for secret updation behavior
